ida_build/lib/xfer/DOC/client.3
.\" $Id: client.3,v 1.1.1.1 2000/02/08 20:20:43 dec Exp $
.TH XFER_CLIENT 3 "June 1996" "IDA Data Exchange Routines"
.SH NAME
Xfer_Connect, Xfer_RecvWav, Xfer_ErrStr, Xfer_Exit \- client side routines for dialog with an IDA data server
.SH SYNOPSIS
.B cc
.RI "[ " "flag" " \|.\|.\|. ] " "file" " \|.\|.\|."
.B \-lxfer \-lutil
.RI "[ " "library" " \|.\|.\|. ]"
.LP
.B #include "xfer.h"
.ft
.fi
.LP
.nf
.ft B
int Xfer_Connect(host, service, port, proto, request, cnf)
char *host;
char *service;
int port;
int proto;
struct xfer_req *request;
struct xfer_cnf *cnf;
.ft
.fi
.LP
.nf
.ft B
int Xfer_RecvWav(sd, wav)
int sd;
struct xfer_wav *wav;
.ft
.fi
.LP
.nf
.ft B
char *Xfer_ErrStr()
.ft
.fi
.LP
.nf
.ft B
void Xfer_Exit(sd, status)
int sd;
int status;
.ft
.fi
.SH DESCRIPTION
These functions support dialog with an IDA data server, following
the \fInrts_xfer2\fR protocol.  The
information provided here assumes familiarity with the protocol and
the implementation specific data structures, described
\fBIDA Data Exchange Protocol\fR document.
.sp
.B Xfer_Connect(\|)
initiates a socket connection with \fIhost\fR, where \fIhost\fR is
a null-terminated string containing the remote hostname or IP address
in the conventional \fB` . '\fR notation.  
The argument
\fIproto\fR determines the communication protocol to be
used.  Currently, this \fImust\fR be the null-terminated string `\fBtcp\fR'.
One of \fIservice\fR or \fIport\fR
is used to determine the port number at which to connect.  
If \fIservice\fR
is not \fBNULL\fR, then it is taken to be the name of a service defined
on the local host, and is used with \fIproto\fR and 
\fIgetservbyname\fR(3N)
to look up the port number.  If \fIservice\fR is \fBNULL\fR then \fIport\fR
is taken to be the port number.  Once the connection is established
then the request message pointed to by \fIrequest\fR is sent to the
server.
If the server accepts the connection and the request, then
it will return the remote configuration which will be stored in the
structure pointed to by \fIcnf\fR.
.sp
All socket I/O is set to time out according to the value of the
\fItimeout\fR field in the \fIrequest\fR structure.  This value
is silently forced to be no less than the
the system wide default of \fBXFER_DEFTO\fR (normally 60 seconds).
.sp
If the request was for a remote configuration, then all dialog
with the server is complete after the function returns.  If the
request was for waveform data, then the program should loop around
calls to \fBXfer_RecvWav(\|)\fR.
.sp
.B Xfer_RecvWav(\|)
retrieves a waveform message from the server connected via the socket
\fIsd\fR, decompressing if necessary, and stores it in the structure pointed to by \fIwav\fR.  
Unless a raw waveform format was requested, the data will be in the 
host's native byte order.
.sp
.B Xfer_ErrStr(\|)
returns a null-terminated string describing the error condition 
reflected by the current value of the global variable \fIxfer_errno\fR.
.sp
\fBXfer_Exit(\|)\fR does a \fIshutdown\fR(2) on the socket \fIsd\fR (if it
is a positive, non-zero descriptor), logs a message indicating the
exit \fIstatus\fR and exits.  If you have the socket descriptor available,
then you should pass it to the routine. Otherwise (if you just want to
get the logged exit) set \fIsd\fR to be zero or negative.
.SH SIGNALS
\fBXfer_Connect(\|)\fR installs signal handlers to trap
\fBSIGPIPE\fR (for writes to a server which has died) and
\fBSIGALRM\fR (to handle socket I/O timeouts).  In addition,
\fBSIGUSR1\fR and \fBSIGUSR2\fR are trapped by the logging facility
if it is enabled (see NOTES).
.SH RETURN VALUES
.B Xfer_Connect(\|)
returns the socket descriptor of the connection if it was successfully
established \fIand\fR the request accepted by the server, and a negative
number in case of error.
.sp
\fBXfer_RecvWav(\|)\fR,
returns either \fBXFER_OK\fR or \fBXFER_FINISHED\fR upon success and
\fBXFER_ERROR\fR in case of failure.  If \fBXFER_FINISHED\fR is returned
then the server has provided all requested data.
.sp
\fBXfer_ErrStr(\|)\fR returns a pointer to a null-terminated string.  The
pointer is to static data whose contents are overwritten by each call.
.sp
\fBXfer_Exit(\|)\fR
does not return.
.sp
In case of failure, all functions set the global variable \fIxfer_errno\fR
to indicate the reason.  The possible values of \fIxfer_errno\fR are
given below.
.SH ERRORS
.TP 18
.SM XFER_EBUSY
The server is too busy to accept the connection.
.TP
.SM XFER_ECONNECT
No connection with peer.
.TP
.SM XFER_EFORMAT
Unsupported format.
.TP
.SM XFER_EHANDLER
Unable to install signal handler.
Refer to the global variable \fIerrno\fR for the cause.
.TP
.SM XFER_EINVAL
Illegal data were received.  
This indicates a bug in the server, should it ever occur
in \fBXfer_Connect\fR() or \fBXfer_RecvWav\fR().
.TP
.SM XFER_EIO
An I/O error occured.  
Refer to the global variable \fIerrno\fR for the cause.
.TP
.SM XFER_ELIMIT
An implementation defined limit was exceeded.
.TP
.SM XFER_EPIPE
The server died.
.TP
.SM XFER_EPROTOCOL
The server does not support the specified protocol version.
.TP
.SM XFER_ENOSUCH
None of the requested stations/channels are available.
.TP
.SM XFER_EREFUSED
The request was sent by an unauthorized client and the server refused
to accept the connection.
.TP
.SM XFER_EREJECTED
The request was rejected by the server, due to invalid contents.
.TP
.SM XFER_ETIMEDOUT
The socket I/O time limit expired.
.TP
.SM XFER_ETOOBIG
An incoming message exceeded the implementation message size limit.
This indicates a bug in the server, should it ever occur.
.SH NOTES
It is possible to enable informational and error messages which are
logged (via \fIutil_log\fR(3)) to a file or to the \fIsyslogd\fR(8)
daemon by calling \fIutil_loginit\fR(3) during program initialization,
if desired.
.sp
The connection with the server is made using the general purpose
function \fIutil_connect\fR(3).
.sp
.SH SEE ALSO
.BR util_loginit (3),
.BR util_log (3)


ida_build/windows/apps/crtd/CLIENT.CPP
/* @(#)client.c	1.18 03/18/97 */
/*======================================================================
 * 
 * IDA data exchange protocols, client side routines.
 *
 * - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
 * Copyright (c) 1997 Regents of the University of California.
 * All rights reserved.
 *====================================================================*/
#include "stdafx.h"
#include <winsock.h>
#include <errno.h>
#include <memory.h>
#include "xfer.h"

long util_ldcmp(long *dest, char *src, long  nsamp);
short util_sdcmp(short *dest, char *src, long  nsamp);
unsigned long util_order();
void util_lswap(long *input, long number);
void util_sswap(short *input, long number);



static int slen = sizeof(u_short);
static int llen = sizeof(u_long);
static u_short sval;
static u_long lval;

/*============================ internal use ==========================*/

/*  Decompress waveform data */

int xfer_Decompress(struct xfer_cnf *cnf, struct xfer_wav *wav)
	{
	unsigned long order;
	static union 
		{
		short s[XFER_MAXDAT];
		long  l[XFER_MAXDAT];
		} buffer;
char *data, *output;
int nsamp, wrdsiz, standx, chnndx;
long *nbyte;
int *comp;
static char *fid = "xfer_Decompress";

/* The following formats don't have protocol compression */

    if (
        wav->format == XFER_WAVRAW    ||
        wav->format == XFER_WAVSEED   ||
        wav->format == XFER_WAVPASSCAL
    ) return XFER_OK;

/* Pull out the necessary pieces from the waveform record */

    switch (wav->format) {

      case XFER_WAVGEN1:

        standx =  wav->type.gen1.standx;
        chnndx =  wav->type.gen1.chnndx;
        nbyte  = &wav->type.gen1.nbyte;
        nsamp  =  wav->type.gen1.nsamp;
        data   =  wav->type.gen1.data;
        comp   = &wav->type.gen1.comp;

        break;


      default:
//        util_log(1, "%s: unsupported waveform format ", fid);
        xfer_errno = XFER_EFORMAT;
        return XFER_ERROR;
    }

/* Pull out the necessary pieces from the configuration record */

    switch (cnf->format) {

      case XFER_CNFGEN1:

        wrdsiz = cnf->type.gen1.sta[standx].chn[chnndx].wrdsiz;
        order  = cnf->type.gen1.sta[standx].chn[chnndx].order;
        break;

      case XFER_CNFNRTS:

        wrdsiz = cnf->type.nrts.sta[standx].chn[chnndx].wrdsiz;
        order  = cnf->type.nrts.sta[standx].chn[chnndx].order;
        break;

      default:

//        util_log(1, "%s: unsupported configuration format code %d",
//            fid, wav->format
//        );
        xfer_errno = XFER_EFORMAT;
        return XFER_ERROR;
    }

/* Sanity checks */

    if (wrdsiz != 2 && wrdsiz != 4) {
//        util_log(1, "%s: unexpected wrdsiz `%d'", fid, wrdsiz);
        xfer_errno = XFER_EINVAL;
        return XFER_ERROR;
    }


/* Decompress */

    switch (*comp) {

      case XFER_CMPIGPP:

        if (wrdsiz == 4) {
            util_ldcmp(buffer.l, data, nsamp);
            output = (char *) buffer.l;
        } else {
            util_sdcmp(buffer.s, data, nsamp);
            output = (char *) buffer.s;
        }
        break;

      case XFER_CMPNONE:

        if (order != util_order()) {
            if (wrdsiz == 4) {
                util_lswap((long *) data, nsamp);
            } else {
                util_sswap((short *) data, nsamp);
            }
        }
        output = data;
        break;

      default:

//        util_log(1, "%s: unrecognized compression code %d", fid, *comp);
        xfer_errno = XFER_EINVAL;
        return XFER_ERROR;
    }

    *comp  = XFER_CMPNONE;
    *nbyte = nsamp * wrdsiz;
    memcpy(data, output, *nbyte);

    return XFER_OK;
}

/* Encode a channel portion of a station portion of a waveform request */

long xfer_EncodeChnReq(char *start_of_message, void *info, int protocol)
{
char *ptr;
struct xfer_time *xtime;
struct xfer01_chnreq *ver01;
static char *fid = "xfer_EncodeChnReq";

    ptr = start_of_message;

    switch (protocol) {

      case 0x01:

        ver01 = (struct xfer01_chnreq *) info;

        memcpy(ptr, ver01->name, XFER01_CNAMLEN + 1);
        ptr += XFER01_CNAMLEN + 1;

        xtime = xfer_time(ver01->beg);
        lval = htonl((u_long) xtime->sec);
        memcpy(ptr, &lval, llen);
        ptr += llen;
        lval = htonl(xtime->usec);
        memcpy(ptr, &lval, llen);
        ptr += llen;

        xtime = xfer_time(ver01->end);
        lval = htonl((u_long) xtime->sec);
        memcpy(ptr, &lval, llen);
        ptr += llen;
        lval = htonl(xtime->usec);
        memcpy(ptr, &lval, llen);
        ptr += llen;

        break;

      default:
        xfer_errno = XFER_EPROTOCOL;
        return XFER_ERROR;
    }

    return (long) (ptr - start_of_message);
}

/* Encode a station portion of a waveform request */

long xfer_EncodeStaReq(char *start_of_message, void *info, int protocol)
{
int i;
long len;
char *ptr;
struct xfer01_stareq *ver01;
static char *fid = "xfer_EncodeStaReq";

    ptr = start_of_message;

    switch (protocol) {

      case 0x01:

        ver01 = (struct xfer01_stareq *) info;

        memcpy(ptr, ver01->name, XFER01_SNAMLEN + 1);
        ptr += XFER01_SNAMLEN + 1;

        sval = htons((u_short) ver01->nchn);
        memcpy(ptr, &sval, slen);
        ptr += slen;

        for (i = 0; i < ver01->nchn; i++) {
            len = xfer_EncodeChnReq(ptr, (void *)(ver01->chn+i), protocol);
            if (len <= 0) return XFER_ERROR;
            ptr += len;
        }

        break;

      default:
        xfer_errno = XFER_EPROTOCOL;
        return XFER_ERROR;
    }

    return (long) (ptr - start_of_message);
}

/* Encode a waveform request */

long xfer_EncodeWavReq(char *start_of_message, union xfer_wavreq *req, int protocol)
{
int i;
long len;
char *ptr;
static char *fid = "xfer_EncodeWavReq";

    ptr = start_of_message;

    switch (protocol) {

      case 0x01:

        sval = htons((u_short) req->ver01.format);
        memcpy(ptr, &sval, slen);
        ptr += slen;

        sval = htons((u_short) req->ver01.keepup);
        memcpy(ptr, &sval, slen);
        ptr += slen;

        sval = htons((u_short) req->ver01.comp);
        memcpy(ptr, &sval, slen);
        ptr += slen;

        sval = htons((u_short) req->ver01.nsta);
        memcpy(ptr, &sval, slen);
        ptr += slen;

        for (i = 0; i < req->ver01.nsta; i++) {
            len = xfer_EncodeStaReq(ptr,(void *)(req->ver01.sta+i),protocol);
            if (len <= 0) return XFER_ERROR;
            ptr += len;
        }

        break;

      default:
        xfer_errno = XFER_EPROTOCOL;
        return XFER_ERROR;
    }

    return (long) (ptr - start_of_message);
}

/* Encode a request preamble */

long xfer_EncodePreamble(char *start_of_message, union xfer_preamble *preamble, int protocol)
{
char *ptr;
static char *fid = "xfer_EncodePreamble";

    ptr = start_of_message;

    switch (protocol) {

      case 0x01:

        lval = htonl((u_long) preamble->ver01.client_id);
        memcpy(ptr, &lval, llen);
        ptr += llen;

        sval = htons((u_short) preamble->ver01.format);
        memcpy(ptr, &sval, slen);
        ptr += slen;

        break;

      default:
        xfer_errno = XFER_EPROTOCOL;
        return XFER_ERROR;
    }

    return (long) (ptr - start_of_message);
}

/* Encode an arbitrary request */

long xfer_EncodeReq(char *start_of_message, struct xfer_req *req)
{
long len;
char *ptr;
static char *fid = "xfer_EncodeReq";

    ptr = start_of_message;

/* ALL requests include the protocol as the first two bytes */

    sval = htons((u_short) req->protocol);
    memcpy(ptr, &sval, slen);
    ptr += slen;

/* AND followed by 4 bytes of request code */

    lval = htonl((u_long) req->type);
    memcpy(ptr, &lval, llen);
    ptr += llen;

/* AND followed by 2 bytes of socket I/O timeout */

    sval = htons((u_short) req->timeout);
    memcpy(ptr, &sval, slen);
    ptr += slen;

/* AND followed by 4 bytes of client send (server recv) buffer size */

    lval = htonl((u_long) req->sndbuf);
    memcpy(ptr, &lval, llen);
    ptr += llen;

/* AND followed by 4 bytes of client recv (server send) buffer size */

    lval = htonl((u_long) req->rcvbuf);
    memcpy(ptr, &lval, llen);
    ptr += llen;

/* Everything else that follows depends on the protocol version */

    switch (req->protocol) {

      case 0x01:

        len = xfer_EncodePreamble(ptr, &req->preamble, req->protocol);
        if (len < 0) return XFER_ERROR;
        ptr += len;

        switch (req->type) {

          case XFER_CNFREQ:
            /* nothing further is required for a configuration request */
            break;

          case XFER_WAVREQ:
            len = xfer_EncodeWavReq(ptr, &req->request.wav, req->protocol);
            if (len < 0) return XFER_ERROR;
            ptr += len;

            break;

          default:
            xfer_errno = XFER_EREQUEST;
            return XFER_ERROR;
        }

        break;

      default:
        xfer_errno = XFER_EPROTOCOL;
        return XFER_ERROR;
    }

    return (long) (ptr - start_of_message);
}

/* Decode a channel descriptor */

int xfer_DecodeChnCnf(void *info, char **src, int format)
{
float factor;
struct xfer_time xtime;
struct xfer_gen1chncnf *gen1;
static char *fid = "xfer_DecodeChnCnf";

    switch (format) {

      case XFER_CNFGEN1:

        gen1 = (struct xfer_gen1chncnf *) info;

        memcpy(gen1->name, *src, XFER01_CNAMLEN + 1);
        *src += XFER01_CNAMLEN + 1;

        memcpy(gen1->instype, *src, XFER01_INAMLEN + 1);
        *src += XFER01_INAMLEN + 1;

        memcpy(&sval, *src, slen);
        *src += slen;
        gen1->wrdsiz = (int) ntohs(sval);

        memcpy(&lval, *src, llen);
        *src += llen;
        gen1->order = (unsigned long) ntohl(lval);

        memcpy(&lval, *src, llen);
        *src += llen;
        gen1->sint = (float) ntohl(lval) / 1000000.0f;

        memcpy(&lval, *src, llen);
        *src += llen;
        factor = (float) ntohl(lval);

        memcpy(&lval, *src, llen);
        *src += llen;
        gen1->calib = (float) ntohl(lval) / factor;

        memcpy(&lval, *src, llen);
        *src += llen;
        gen1->calper = (float) ntohl(lval) / 100000.0f;

        memcpy(&lval, *src, llen);
        *src += llen;
        gen1->vang = (float) ntohl(lval) / 1000000.0f;

        memcpy(&lval, *src, llen);
        *src += llen;
        gen1->hang = (float) ntohl(lval) / 1000000.0f;

        memcpy(&lval, *src, llen);
        *src += llen;
        xtime.sec = (long) ntohl(lval);

        memcpy(&lval, *src, llen);
        *src += llen;
        xtime.usec = (u_long) ntohl(lval);
        gen1->beg = xfer_dtime(&xtime);

        memcpy(&lval, *src, llen);
        *src += llen;
        xtime.sec = (long) ntohl(lval);

        memcpy(&lval, *src, llen);
        *src += llen;
        xtime.usec = (u_long) ntohl(lval);
        gen1->end = xfer_dtime(&xtime);

        break;

      default:
        xfer_errno = XFER_EFORMAT;
        return XFER_ERROR;
    }

    return XFER_OK;
}

/* Decode a station descriptor */

int xfer_DecodeStaCnf(void *info, char **src, int format)
{
int i, retval;
struct xfer_gen1stacnf *gen1;
static char *fid = "xfer_DecodeStaCnf";

    switch (format) {

      case XFER_CNFGEN1:

        gen1 = (struct xfer_gen1stacnf *) info;

        memcpy(gen1->name, *src, XFER01_SNAMLEN + 1);
        *src += XFER01_SNAMLEN + 1;

        memcpy(&lval, *src, llen);
        *src += llen;
        gen1->lat = (float) ntohl(lval) / 1000000.0f;

        memcpy(&lval, *src, llen);
        *src += llen;
        gen1->lon = (float) ntohl(lval) / 1000000.0f;

        memcpy(&lval, *src, llen);
        *src += llen;
        gen1->elev = (float) ntohl(lval) / 1000000.0f;

        memcpy(&lval, *src, llen);
        *src += llen;
        gen1->depth = (float) ntohl(lval) / 1000000.0f;

        memcpy(&sval, *src, slen);
        *src += slen;
        gen1->nchn = (int) ntohs(sval);

        if (gen1->nchn > XFER_MAXCHN) {
            xfer_errno = XFER_ELIMIT;
            return XFER_ERROR;
        }

        for (i = 0; i < gen1->nchn; i++) {
            retval = xfer_DecodeChnCnf(gen1->chn + i, src, format);
            if (retval != XFER_OK) return retval;
        }

        break;


      default:
        xfer_errno = XFER_EFORMAT;
        return XFER_ERROR;
    }

    return XFER_OK;
}

/* Decode a configuration message */

int xfer_DecodeCnf(struct xfer_cnf *cnf, char *src)
{
int i, retval;
static char *fid = "xfer_DecodeCnf";

    memcpy(&sval, src, slen);
    src += slen;
    cnf->format = (int) ntohs(sval);

    switch (cnf->format) {

      case XFER_CNFGEN1:

        memcpy(&lval, src, llen);
        src += llen;
        cnf->type.gen1.order = (u_long) ntohl(lval);

        memcpy(&sval, src, slen);
        src += slen;
        cnf->type.gen1.nsta = (int) ntohs(sval);

        if (cnf->type.gen1.nsta > XFER_MAXSTA) {
/*            util_log(1, "%s: increase XFER_MAXSTA (%d) to %d",
                fid, XFER_MAXSTA, cnf->type.gen1.nsta
            );*/
            xfer_errno = XFER_ELIMIT;
            return XFER_ERROR;
        }

        for (i = 0; i < cnf->type.gen1.nsta; i++) {
            retval = xfer_DecodeStaCnf(
                cnf->type.gen1.sta + i, &src, cnf->format
            );
            if (retval != XFER_OK) return retval;
        }

        break;

#ifdef NRTS_SUPPORT
      case XFER_CNFNRTS:

        memcpy(&lval, src, llen);
        src += llen;
        cnf->type.nrts.order = (u_long) ntohl(lval);

        memcpy(&sval, src, slen);
        src += slen;
        cnf->type.nrts.nsta = (int) ntohs(sval);

        if (cnf->type.nrts.nsta > XFER_MAXSTA) {
            util_log(1, "%s: increase XFER_MAXSTA (%d) to %d",
                fid, XFER_MAXSTA, cnf->type.nrts.nsta
            );
            xfer_errno = XFER_ELIMIT;
            return XFER_ERROR;
        }

        for (i = 0; i < cnf->type.nrts.nsta; i++) {
            retval = xfer_DecodeStaCnf(
                cnf->type.nrts.sta + i, &src, cnf->format
            );
            if (retval != XFER_OK) return retval;
        }

        break;
#endif /* NRTS_SUPPORT */

      default:
        xfer_errno = XFER_EFORMAT;
        return XFER_ERROR;
    }

    return XFER_OK;
}

/* Receive a remote configuration */

int xfer_RecvCnf
(int sd,struct xfer_cnf *cnf)
{
int retval;
long len;
static char *fid = "xfer_RecvCnf";

   retval = xfer_RecvMsg(sd, Xfer_Buffer, Xfer_Buflen, &len);
   if (retval != XFER_OK) return XFER_ERROR;

/* If a break (zero length message) was received, follow up with
 * a second read to get the cause code.
 */

    if (len == 0) {
        retval = xfer_RecvMsg(sd, Xfer_Buffer, Xfer_Buflen, &len);
        if (retval != XFER_OK) return XFER_ERROR;
        memcpy(&lval, Xfer_Buffer, llen);
        xfer_errno = (int) ntohl(lval);
        return XFER_ERROR;
    }

/* Otherwise decode the presumably OK configuration message */

    return xfer_DecodeCnf(cnf, Xfer_Buffer);
}

/* Decode a waveform packet */

int xfer_DecodeWav(struct xfer_cnf *cnf,struct xfer_wav *wav, char *src)
{
struct xfer_time xtime;
static char *fid = "xfer_DecodeWav";

    memcpy(&sval, src, slen);
    src += slen;
    wav->format = (int) ntohs(sval);

    switch (wav->format) {

      case XFER_WAVGEN1:

        memcpy(&sval, src, slen);
        src += slen;
        wav->type.gen1.standx = (int) ntohs(sval);

        memcpy(&sval, src, slen);
        src += slen;
        wav->type.gen1.chnndx = (int) ntohs(sval);

        memcpy(&lval, src, llen);
        src += llen;
        xtime.sec = (long) ntohl(lval);

        memcpy(&lval, src, llen);
        src += llen;
        xtime.usec = (u_long) ntohl(lval);
        wav->type.gen1.tofs = xfer_dtime(&xtime);

        memcpy(&sval, src, slen);
        src += slen;
        wav->type.gen1.tear = (int) ntohs(sval);

        memcpy(&sval, src, slen);
        src += slen;
        wav->type.gen1.comp = (int) ntohs(sval);

        memcpy(&lval, src, llen);
        src += llen;
        wav->type.gen1.nsamp = (long) ntohl(lval);

        memcpy(&lval, src, llen);
        src += llen;
        wav->type.gen1.nbyte = (long) ntohl(lval);

        if (wav->type.gen1.nbyte > XFER_MAXDAT) {
/*            util_log(1, "%s: increase XFER_MAXDAT (%d) to %d",
                fid, XFER_MAXDAT, wav->type.gen1.nbyte
            );*/
            xfer_errno = XFER_ELIMIT;
            return XFER_ERROR;
        }

        if (wav->type.gen1.nbyte < 0) {
/*            util_log(1, "%s: illegal gen1.nbyte %d",
                fid, wav->type.gen1.nbyte
            );*/
            xfer_errno = XFER_EINVAL;
            return XFER_ERROR;
        }

        memcpy(wav->type.gen1.data, src, wav->type.gen1.nbyte);

        break;

      case XFER_WAVRAW:

        memcpy(&lval, src, llen);
        src += llen;
        wav->type.raw.nbyte = (long) ntohl(lval);

        if (wav->type.raw.nbyte > XFER_MAXDAT) {
/*            util_log(1, "%s: increase XFER_MAXDAT (%d) to %d",
                fid, XFER_MAXDAT, wav->type.raw.nbyte
            );*/
            xfer_errno = XFER_ELIMIT;
            return XFER_ERROR;
        }

        if (wav->type.raw.nbyte < 0) {
/*            util_log(1, "%s: illegal raw.nbyte %d",
                fid, wav->type.raw.nbyte
            );*/
            xfer_errno = XFER_EINVAL;
            return XFER_ERROR;
        }

        memcpy(wav->type.raw.data, src, wav->type.raw.nbyte);

        break;

      default:
/*        util_log(1, "%s: unsupported waveform format code %d",
            fid, wav->format
        );*/
        xfer_errno = XFER_EFORMAT;
        return XFER_ERROR;
    }

    return xfer_Decompress(cnf, wav);
}

/* Send an arbitrary request */

int xfer_SendReq(int sd,struct xfer_req *req)
{
long len;
static char *fid = "xfer_SendReq";

    len = xfer_EncodeReq(Xfer_Buffer+4, req);
    if (len < 0) return XFER_ERROR;

    return xfer_SendMsg(sd, Xfer_Buffer, len);
}

/*============================ external use ==========================*/

/* Receive a waveform packet */

int Xfer_RecvWav(int sd, struct xfer_cnf *cnf, struct xfer_wav *wav)
{
int retval = XFER_OK;
long len;
static char *fid = "Xfer_RecvWav";

    while (1) {

        retval = xfer_RecvMsg(sd, Xfer_Buffer, Xfer_Buflen, &len);
        if (retval != XFER_OK) return XFER_ERROR;

        if (len > 0) return xfer_DecodeWav(cnf, wav, Xfer_Buffer);

    /* If a break (zero length message) was received, follow up with
     * a second read to get the cause code.
     */
        retval = xfer_RecvMsg(sd, Xfer_Buffer, Xfer_Buflen, &len);
        if (retval != XFER_OK) return XFER_ERROR;
        memcpy(&lval, Xfer_Buffer, llen);
        retval = (int) ntohl(lval);

        if (retval == XFER_FINISHED) {
            xfer_errno = 0;
            return retval;
        } else if (retval != XFER_HEARTBEAT) {
            xfer_errno = retval;
            return XFER_ERROR;
        } else {
//            util_log(2, "HEARTBEAT received");
        }
    }
}
